include/shark/Algorithms/DirectSearch/Operators/Hypervolume/HypervolumeSubsetSelection2D.h

```diff
@@ -0,0 +1,298 @@
+ * 
+ * Shark is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published 
+ * by the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ * 
+ * Shark is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ * 
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Shark.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+#ifndef SHARK_ALGORITHMS_DIRECTSEARCH_HYPERVOLUMESUBSETSELECTION_2D_H
+#define SHARK_ALGORITHMS_DIRECTSEARCH_HYPERVOLUMESUBSETSELECTION_2D_H
+
+#include <shark/LinAlg/Base.h>
+
+#include <algorithm>
+#include <vector>
+#include <deque>
+
+namespace shark {
+/// \brief Implementation of the exact hypervolume subset selection algorithm in 2 dimensions.
+///
+/// This algorithm solves the problem of selecting a subset of points with largest hypervolume in 2D.
+/// The algorithm has complexity n (k+log(n)) 
+///
+/// The algorithm is described in:
+/// Bringmann, Karl, Tobias Friedrich, and Patrick Klitzke. 
+/// "Two-dimensional subset selection for hypervolume and epsilon-indicator."
+/// Proceedings of the 2014 conference on Genetic and evolutionary computation. 
+/// ACM, 2014.	
+/// (although it is not very helpful)
+struct HypervolumeSubsetSelection2D {
+private:
+	
+	struct Point{
+		Point(){}
+
+		Point(double f1, double f2, std::size_t index)
+		: f1(f1)
+		, f2(f2)
+		, index(index)
+		, selected(false)
+		{}
+			
+		bool operator<(Point const& rhs) const{//for lexicographic sorting
+			if (f1 < rhs.f1) return true;
+			if (f1 > rhs.f1) return false;
+			return (f2 < rhs.f1);
+		}
+		
+		double f1;
+		double f2;
+		std::size_t index;
+		bool selected;
+	};
+	
+	///\brief Linear function a*x+b where a is stored in first and b is stored in section.
+	///
+	/// The linear function also stores an index to uniquely identify it
+	struct LinearFunction{
+	
+		double a;
+		double b;
+		std::size_t index;
+		
+		LinearFunction(double a, double b, std::size_t index = 0):a(a), b(b), index(index){}
+		LinearFunction(){}
+		
+		double eval(double x)const{
+			return a*x + b;
+		}
+	};
+
+	/// \brief Returns the intersection of two linear functions
+	double Intersection(LinearFunction f1, LinearFunction f2)const{
+		return (f2.b - f1.b) / (f1.a - f2.a);
+	}
+	
+	
+	/// \brief  Calculates for each given x the maximum among the functions f, i.e. the upper envelope of f.
+	/// 
+	/// Algoruthm 2 in the paper. Complexity O(n)
+	/// given a set of functions f_1...f_n, ordered by slope such that f_1.a < f_2.a<... and points with x-coordinate x_1<...<x_n
+	/// computes h_i = max_{1 <= j < i} f_j(x_i) for i=1,...,n as well as the index of the function leading to the value h_i
```

Comment:
This description suggets that h_1 is undefined since the index set 1 <= j < i is empty for i=1. However, the implementation seems to use the index set 1 <= j <= i.